src/main/java/cam72cam/immersiverailroading/gui/PaintBrushPicker.java

```diff
@@ -0,0 +1,207 @@
+import cam72cam.mod.entity.Entity;
+import cam72cam.mod.entity.Player;
+import cam72cam.mod.gui.helpers.GUIHelpers;
+import cam72cam.mod.gui.screen.Button;
+import cam72cam.mod.gui.screen.IScreen;
+import cam72cam.mod.gui.screen.IScreenBuilder;
+import cam72cam.mod.gui.screen.TextField;
+import cam72cam.mod.render.opengl.RenderState;
+import util.Matrix4;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class PaintBrushPicker implements IScreen {
+    private static final Random r = new Random();
+
+    private EntityMoveableRollingStock stock;
+    private String variant;
+
+    private List<Button> options;
+
+    private int page;
+    private int pageSize;
+    private Button pagination;
+    private long frame;
+
+    @Override
+    public void init(IScreenBuilder screen) {
+        frame = 0;
+
+        Entity ent = MinecraftClient.getEntityMouseOver();
+        if (ent == null) {
+            screen.close();
+            return;
+        }
+        this.stock = ent.as(EntityMoveableRollingStock.class);
+        if (stock == null) {
+            screen.close();
+        }
+        this.variant = stock.getTexture();
+
+        int xtop = -GUIHelpers.getScreenWidth() / 2;
+        int ytop = -GUIHelpers.getScreenHeight()/4;
+        int width = 200;
+        int height = 20;
+        page = 0;
+        pageSize = Math.max(1, GUIHelpers.getScreenHeight() / height - 2);
+
+        TextField search = new TextField(screen, xtop + 1, ytop + 1, width - 2, height - 2);
+
+        pagination = new Button(screen, xtop, ytop + height, width + 1, height, "Page") {
+            @Override
+            public void onClick(Player.Hand hand) {
+                page += hand == Player.Hand.PRIMARY ? 1 : -1;
+                updateVariants(search.getText());
+            }
+        };
+
+        options = new ArrayList<>();
+        for (int i = 0; i < pageSize; i++) {
+            options.add(new Button(screen, xtop, ytop + height*2 + i * height, width+1, height, "") {
+                @Override
+                public void onClick(Player.Hand hand) {
+                    variant = stock.getDefinition().textureNames.entrySet().stream()
+                            .filter(x -> x.getValue().equals(this.getText()))
+                            .map(Map.Entry::getKey)
+                            .findFirst().orElse(null);
+                }
+            });
+        }
+
+        search.setValidator(s -> {
+            page = 0;
+            this.updateVariants(s);
+            return true;
+        });
+        this.updateVariants("");
+
+        width = 60;
+        Button random = new Button(screen, GUIHelpers.getScreenWidth() / 2 - width, ytop, width, height, "Random") {
+            @Override
+            public void onClick(Player.Hand hand) {
+                List<String> choices = new ArrayList<>(stock.getDefinition().textureNames.keySet());
+                variant = choices.get(r.nextInt(choices.size()));
```

Comment:
There's a regression here. It's possible to get the same variant twice in a row, which is both a confusing UX and an annoyance for stock with only a few liveries.